crates/sui-graphql-rpc/src/types/transaction_block/tx_lookups.rs

```diff
@@ -0,0 +1,468 @@
+            lo,
+            hi,
+            page.after().map(|x| x.tx_sequence_number),
+            page.before().map(|x| x.tx_sequence_number),
+            scan_limit,
+            page.is_from_front(),
+        )))
+    }
+
+    pub(crate) fn tx_lo(&self) -> u64 {
+        max_option([self.after, Some(self.lo)]).unwrap()
+    }
+
+    pub(crate) fn tx_hi(&self) -> u64 {
+        min_option([self.before, Some(self.hi)]).unwrap()
+    }
+
+    /// The lower bound `tx_sequence_number`` of the range to scan within. This defaults to the min
+    /// tx_sequence_number of the checkpoint bound. If a cursor is provided, the lower bound is
+    /// adjusted to the larger of the two. The resulting value is additionally modified by the
+    /// `scan_limit` if `is_from_front` is false.
+    pub(crate) fn scan_lo(&self) -> u64 {
+        let adjusted_lo = self.tx_lo();
+
+        if self.is_from_front {
+            adjusted_lo
+        } else {
+            // If not from the front, then the scan_limit must only be applied to the lower bound
+            if let Some(scan_limit) = self.scan_limit {
+                adjusted_lo.max(self.tx_hi().saturating_sub(scan_limit))
+            } else {
+                adjusted_lo
+            }
+        }
+    }
+
+    /// The upper bound `tx_sequence_number` of the range to scan within. This defaults to the max
+    /// tx_sequence_number of the checkpoint bound. If a cursor is provided, the upper bound is
+    /// adjusted to the smaller of the two. The resulting value is additionally modified by the
+    /// `scan_limit` if `is_from_front` is true.
+    pub(crate) fn scan_hi(&self) -> u64 {
+        let adjusted_hi = self.tx_hi();
+
+        if self.is_from_front {
+            if let Some(scan_limit) = self.scan_limit {
+                adjusted_hi.min(self.tx_lo().saturating_add(scan_limit))
+            } else {
+                adjusted_hi
+            }
+        } else {
+            adjusted_hi
+        }
+    }
+
+    /// If the query result does not have a previous page, check whether the scan limit is within
+    /// the initial tx_sequence_number range.
+    pub(crate) fn scan_has_prev_page(&self) -> bool {
+        if self.after.unwrap_or(0) >= self.hi {
+            return false;
+        }
+
+        self.scan_lo() > self.lo
+    }
+
+    /// If the query result does not have a next page, check whether the scan limit is within the
+    /// initial tx_sequence_number range.
+    pub(crate) fn scan_has_next_page(&self) -> bool {
+        if self.before.unwrap_or(self.hi) <= self.lo {
+            return false;
+        }
+
+        self.scan_hi() < self.hi
+    }
+
+    pub(crate) fn scan_prev_page_and_lo(&self) -> (bool, u64) {
+        let has_prev = self.scan_has_prev_page();
+
+        (has_prev, self.scan_lo())
+    }
+
+    pub(crate) fn scan_next_page_and_hi(&self) -> (bool, u64) {
+        let has_next = self.scan_has_next_page();
+
+        (has_next, self.scan_hi())
+    }
```

Comment:
nit: (although this makes me wonder if it's all that useful to have this function?)
```suggestion
        (self.scan_has_prev_page(), self.scan_lo())
    }

    pub(crate) fn scan_next_page_and_hi(&self) -> (bool, u64) {
        (self.scan_has_next_page(), self.scan_hi())
    }
```

Reply:
i suppose not anymore